In [1]:
# Load packages
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
test1 = pd.read_csv('/data/home/stateun/results/test1_results.csv')
test2 = pd.read_csv('/data/home/stateun/results/test2_results.csv')
test3 = pd.read_csv('/data/home/stateun/results/test3_results.csv')
test4 = pd.read_csv('/data/home/stateun/results/test4_results.csv')
test5 = pd.read_csv('/data/home/stateun/results/test5_results.csv')

In [3]:
test = pd.concat([test1, test2, test3, test4, test5], axis = 0, ignore_index = True)
test = test.drop(['dataset'], axis = 1)

In [4]:
test

model    AUC    ACC  Precision  Recall  F1_score  \
0                 IGNnet  79.55  89.62      56.30   56.54     56.41   
1     LogisticRegression  78.84  93.51      59.44   50.83     50.25   
2           DecisionTree  52.59  86.77      51.90   52.59     52.08   
3           RandomForest  80.37  93.06      55.77   51.08     50.85   
4                    MLP  80.51  91.89      57.93   54.35     55.35   
...                  ...    ...    ...        ...     ...       ...   
3623        DecisionTree  63.02  90.66      60.40   63.02     61.50   
3624        RandomForest  80.34  93.64      58.36   51.79     52.13   
3625                 MLP  79.04  91.83      58.10   56.17     56.94   
3626                 SVM  74.48  94.29      47.15   50.00     48.53   
3627             XGBoost  82.23  92.74      53.47   51.31     51.44   

                                            best_params  
0     alpha:0.01, sampling:smote, batch_size:64, epo...  
1     {'C': 1.0, 'class_weight': None, 'dual': False...  
2     {'ccp_alpha': 0.0, 'class_weight': None, 'crit...  
3     {'bootstrap': True, 'ccp_alpha': 0.0, 'class_w...  
4     {'activation': 'relu', 'alpha': 0.0001, 'batch...  
...                                                 ...  
3623  {'ccp_alpha': 0.0, 'class_weight': None, 'crit...  
3624  {'bootstrap': True, 'ccp_alpha': 0.0, 'class_w...  
3625  {'activation': 'relu', 'alpha': 0.0001, 'batch...  
3626  {'C': 1.0, 'break_ties': False, 'cache_size': ...  
3627  {'objective': 'binary:logistic', 'base_score':...  

[3628 rows x 7 columns]

In [5]:
model_list = test['model'].unique()

In [6]:
IGNNet = test[test['model'] == 'IGNnet']
Table = IGNNet[(IGNNet['ACC'] >= 88) & (IGNNet['F1_score'] >= 60)].sort_values(by=['AUC', 'F1_score'], ascending=[False, False])
Table

model    AUC    ACC  Precision  Recall  F1_score  \
3089  IGNnet  83.43  89.19      59.53   61.34     60.32   
2802  IGNnet  82.54  88.65      58.90   62.83     60.31   
2312  IGNnet  82.05  89.23      59.76   61.67     60.58   
1848  IGNnet  82.00  90.60      60.17   60.46     60.31   
1878  IGNnet  82.00  90.60      60.17   60.46     60.31   
1892  IGNnet  82.00  90.60      60.17   60.46     60.31   
1899  IGNnet  82.00  90.60      60.17   60.46     60.31   
1913  IGNnet  82.00  90.60      60.17   60.46     60.31   
2032  IGNnet  82.00  90.60      60.17   60.46     60.31   
2060  IGNnet  82.00  90.60      60.17   60.46     60.31   
2137  IGNnet  82.00  90.60      60.17   60.46     60.31   
2564  IGNnet  80.99  89.36      60.27   62.34     61.16   

                                            best_params  
3089  alpha:0.03, sampling:borderline-smote, batch_s...  
2802  alpha:0.01, sampling:borderline-smote, batch_s...  
2312  alpha:0.03, sampling:smote, batch_size:64, epo...  
1848  alpha:0.01, sampling:smote, batch_size:64, epo...  
1878  alpha:0.01, sampling:smote, batch_size:64, epo...  
1892  alpha:0.01, sampling:smote, batch_size:64, epo...  
1899  alpha:0.01, sampling:smote, batch_size:64, epo...  
1913  alpha:0.01, sampling:smote, batch_size:64, epo...  
2032  alpha:0.01, sampling:smote, batch_size:64, epo...  
2060  alpha:0.01, sampling:smote, batch_size:64, epo...  
2137  alpha:0.01, sampling:smote, batch_size:64, epo...  
2564  alpha:0.05, sampling:smote, batch_size:64, epo...

##### We have selected models with indices 3089 and 2312 as our best models.
##### Therefore, we will consider theri best hyperparameters as our optimal hyperparameters.
##### Plus, we can see that the SMOTE-based models perform better than the other models.

In [7]:
index1 = Table.index[0] - (test1.shape[0] + test2.shape[0] + test3.shape[0] + test4.shape[0] )
index1

1001

In [8]:
index2 = Table.index[2] - (test1.shape[0] + test2.shape[0] + test3.shape[0] + test4.shape[0] )
index2

224

In [9]:
length = len(model_list)
length

7

In [ ]:
best_table1 = test5.iloc[index1 : index1 + length ].drop(['dataset'], axis = 1)
best_table1.drop

model    AUC    ACC  Precision  Recall  F1_score  \
1001              IGNnet  83.43  89.19      59.53   61.34     60.32   
1002  LogisticRegression  79.20  93.17      63.41   51.11     50.64   
1003        DecisionTree  57.19  87.64      55.66   57.19     56.26   
1004        RandomForest  81.29  92.61      59.67   52.32     52.83   
1005                 MLP  78.96  91.14      55.73   53.04     53.68   
1006                 SVM  78.32  93.30      46.67   49.98     48.27   
1007             XGBoost  84.30  92.39      62.14   54.62     56.10   

                                            best_params  
1001  alpha:0.03, sampling:borderline-smote, batch_s...  
1002  {'C': 1.0, 'class_weight': None, 'dual': False...  
1003  {'ccp_alpha': 0.0, 'class_weight': None, 'crit...  
1004  {'bootstrap': True, 'ccp_alpha': 0.0, 'class_w...  
1005  {'activation': 'relu', 'alpha': 0.0001, 'batch...  
1006  {'C': 1.0, 'break_ties': False, 'cache_size': ...  
1007  {'objective': 'binary:logistic', 'base_score':...

In [24]:
best_table1['best_params']

1001    alpha:0.03, sampling:borderline-smote, batch_s...
1002    {'C': 1.0, 'class_weight': None, 'dual': False...
1003    {'ccp_alpha': 0.0, 'class_weight': None, 'crit...
1004    {'bootstrap': True, 'ccp_alpha': 0.0, 'class_w...
1005    {'activation': 'relu', 'alpha': 0.0001, 'batch...
1006    {'C': 1.0, 'break_ties': False, 'cache_size': ...
1007    {'objective': 'binary:logistic', 'base_score':...
Name: best_params, dtype: object

In [11]:
best_table2 = test5.iloc[index2 : index2 + length ].drop(['dataset'], axis = 1)
best_table2

model    AUC    ACC  Precision  Recall  F1_score  \
224              IGNnet  82.05  89.23      59.76   61.67     60.58   
225  LogisticRegression  79.31  93.17      59.20   50.51     49.46   
226        DecisionTree  56.45  87.94      55.40   56.45     55.84   
227        RandomForest  81.77  92.56      57.72   51.69     51.83   
228                 MLP  77.60  91.18      58.42   55.17     56.21   
229                 SVM  78.33  93.34      46.67   50.00     48.28   
230             XGBoost  83.73  92.35      61.49   54.29     55.65   

                                           best_params  
224  alpha:0.03, sampling:smote, batch_size:64, epo...  
225  {'C': 1.0, 'class_weight': None, 'dual': False...  
226  {'ccp_alpha': 0.0, 'class_weight': None, 'crit...  
227  {'bootstrap': True, 'ccp_alpha': 0.0, 'class_w...  
228  {'activation': 'relu', 'alpha': 0.0001, 'batch...  
229  {'C': 1.0, 'break_ties': False, 'cache_size': ...  
230  {'objective': 'binary:logistic', 'base_score':...

In [25]:
best_table2['best_params']

224    alpha:0.03, sampling:smote, batch_size:64, epo...
225    {'C': 1.0, 'class_weight': None, 'dual': False...
226    {'ccp_alpha': 0.0, 'class_weight': None, 'crit...
227    {'bootstrap': True, 'ccp_alpha': 0.0, 'class_w...
228    {'activation': 'relu', 'alpha': 0.0001, 'batch...
229    {'C': 1.0, 'break_ties': False, 'cache_size': ...
230    {'objective': 'binary:logistic', 'base_score':...
Name: best_params, dtype: object

##### Now, We compare the performance between methods about oversamplings

#### ALL Performance

In [13]:
sampling_methods = ['adasyn', 'over-random']

In [14]:
IGNNet['sampling_method'] = IGNNet['best_params'].apply(lambda x: next((s for s in sampling_methods if s in str(x)), None))

best_auc_idx = IGNNet.groupby("sampling_method")["AUC"].idxmax()
best_f1_idx = IGNNet.groupby("sampling_method")["F1_score"].idxmax()
best_auc = IGNNet.loc[best_auc_idx]
best_f1 = IGNNet.loc[best_f1_idx]

merged_perform_all = pd.concat([best_auc, best_f1]).drop_duplicates().reset_index(drop=True)

final_rows = []
for method in sampling_methods:
    subset = merged_perform_all[merged_perform_all['sampling_method'] == method]

    if len(subset) > 1:
        row1, row2 = subset.iloc[0], subset.iloc[1]
        if row1["AUC"] > row2["AUC"] and row1["F1_score"] > row2["F1_score"]:
            final_rows.append(row1)
        elif row2["AUC"] > row1["AUC"] and row2["F1_score"] > row1["F1_score"]:
            final_rows.append(row2)
        else:
            subset = subset.copy() 
            subset["mean_score"] = (subset["AUC"] + subset["F1_score"]) / 2
            chosen_row = subset.sort_values("mean_score", ascending=False).iloc[0]
            final_rows.append(chosen_row)
    elif len(subset) == 1:
        final_rows.append(subset.iloc[0])

final_perform_all = pd.DataFrame(final_rows).reset_index(drop=True).drop(['best_params', 'mean_score'], axis = 1)

/tmp/ipykernel_6444/3800498670.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  IGNNet['sampling_method'] = IGNNet['best_params'].apply(lambda x: next((s for s in sampling_methods if s in str(x)), None))


In [15]:
best_table1

model    AUC    ACC  Precision  Recall  F1_score  \
1001              IGNnet  83.43  89.19      59.53   61.34     60.32   
1002  LogisticRegression  79.20  93.17      63.41   51.11     50.64   
1003        DecisionTree  57.19  87.64      55.66   57.19     56.26   
1004        RandomForest  81.29  92.61      59.67   52.32     52.83   
1005                 MLP  78.96  91.14      55.73   53.04     53.68   
1006                 SVM  78.32  93.30      46.67   49.98     48.27   
1007             XGBoost  84.30  92.39      62.14   54.62     56.10   

                                            best_params  
1001  alpha:0.03, sampling:borderline-smote, batch_s...  
1002  {'C': 1.0, 'class_weight': None, 'dual': False...  
1003  {'ccp_alpha': 0.0, 'class_weight': None, 'crit...  
1004  {'bootstrap': True, 'ccp_alpha': 0.0, 'class_w...  
1005  {'activation': 'relu', 'alpha': 0.0001, 'batch...  
1006  {'C': 1.0, 'break_ties': False, 'cache_size': ...  
1007  {'objective': 'binary:logistic', 'base_score':...

In [16]:
best_table1['sampling_method'] = 'borderline-smote'

In [17]:
best_table2['sampling_method'] = 'smote'

In [22]:
best_table1.iloc[[0], :]
best_table2.iloc[[0], :]

model    AUC    ACC  Precision  Recall  F1_score  \
224  IGNnet  82.05  89.23      59.76   61.67     60.58   

                                           best_params sampling_method  
224  alpha:0.03, sampling:smote, batch_size:64, epo...           smote

In [19]:
final_perform_all

model    AUC    ACC  Precision  Recall  F1_score sampling_method
0  IGNnet  80.31  84.14      57.75   66.33     59.33          adasyn
1  IGNnet  81.60  80.84      57.73   70.28     58.74     over-random

In [23]:
pd.concat([best_table1.iloc[[0], :], best_table2.iloc[[0], :], final_perform_all], axis = 0, ignore_index = True).drop(['best_params'], axis = 1)

model    AUC    ACC  Precision  Recall  F1_score   sampling_method
0  IGNnet  83.43  89.19      59.53   61.34     60.32  borderline-smote
1  IGNnet  82.05  89.23      59.76   61.67     60.58             smote
2  IGNnet  80.31  84.14      57.75   66.33     59.33            adasyn
3  IGNnet  81.60  80.84      57.73   70.28     58.74       over-random